File: src/main/java/com/github/unchama/seichiassist/data/BreakArea.java
package com.github.unchama.seichiassist.data;


import java.util.ArrayList;
import java.util.List;

import org.bukkit.entity.Player;

import com.github.unchama.seichiassist.ActiveSkill;
import com.github.unchama.seichiassist.util.BreakUtil;

public class BreakArea {
	//スキルタイプ番号
	int type;
	//スキルレベル
	int level;
	//フラグ
	int mineflagnum;
	//南向きを基準として破壊の範囲座標
	Coordinate breaklength;
	//破壊回数
	int breaknum;
	//向いている方角
	String dir;
	//破壊範囲を示す相対座標リスト
	List<Coordinate> startlist,endlist;
	//変数として利用する相対座標
	private Coordinate start,end;
	//アサルトスキルの時true
	boolean assaultflag;


	public BreakArea(Player player,int type, int skilllevel,int mineflagnum,boolean assaultflag) {
		this.type = type;
		this.level = skilllevel;
		this.mineflagnum = mineflagnum;
		this.assaultflag = assaultflag;
		this.dir = BreakUtil.getCardinalDirection(player);
		this.startlist = new ArrayList<Coordinate>();
		this.endlist = new ArrayList<Coordinate>();
		//初期化
		ActiveSkill[] as = ActiveSkill.values();
		this.breaklength = as[type-1].getBreakLength(level);
		this.breaknum = as[type-1].getRepeatTimes(level);

		//初期範囲設定
		makeArea();
	}

	public List<Coordinate> getStartList() {
		return startlist;
	}
	public List<Coordinate> getEndList() {
		return endlist;
	}
	public String getDir() {
		return dir;
	}
	public void setDir(String dir) {
		this.dir = new String(dir);
	}
	//破壊範囲の設定
	public void makeArea() {
		startlist.clear();
		endlist.clear();
		//中心座標(0,0,0)のスタートとエンドを仮取得
		start = new Coordinate(-(breaklength.x-1)/2, -(breaklength.y-1)/2, -(breaklength.z-1)/2);
		end = new Coordinate((breaklength.x-1)/2, (breaklength.y-1)/2, (breaklength.z-1)/2);
		//アサルトスキルの時
		if(assaultflag){
			if(type == 6 && level == 10){
				shift(0, (breaklength.y-1)/2 - 1,0);
			}
		}
		//上向きまたは下向きの時
		else if(dir.equals("U") || dir.equals("D")){
			if(!assaultflag && level < 3){}
			else{shift(0, (breaklength.y-1)/2, 0);}
		}
		//それ以外の範囲
		else{
			shift(0, (breaklength.y-1)/2 - 1, (breaklength.z-1)/2);

		}

		if(type == ActiveSkill.BREAK.gettypenum() && level < 3){
			end.add(0, 1, 0);
		}
		if(type == ActiveSkill.BREAK.gettypenum() && level < 3 && mineflagnum == 1){
			shift(0,1,0);
		}

		//スタートリストに追加
		startlist.add(new Coordinate(start));
		endlist.add(new Coordinate(end));

		//破壊回数だけリストに追加

		for(int count = 1; count < breaknum;count++){
			switch(dir){
			case "N":
			case "E":
			case "S":
			case "W":
				shift(0, 0, breaklength.z);
				break;
			case "U":
			case "D":
				if(!assaultflag && level < 3){}
				else{shift(0, breaklength.y, 0);}
				break;

			}
			startlist.add(new Coordinate(start));
			endlist.add(new Coordinate(end));
		}


		switch(dir){
		case "N":
			rotateXZ(180);
			break;
		case "E":
			rotateXZ(270);
			break;
		case "S":
			break;
		case "W":
			rotateXZ(90);
			break;
		case "U":
			break;
		case "D":
			if(!assaultflag)multiply_Y(-1);
			break;
		}
	}
	private void multiply_Y(int i) {
		for(int count = 0;count < breaknum ; count++){
			Coordinate start = startlist.get(count);
			Coordinate end = endlist.get(count);
			Coordinate tmpstart = new Coordinate(startlist.get(count));
			if(i >=0){
				start.setXYZ(start.x,start.y * i,start.z);
				end.setXYZ(end.x,end.y * i,end.z);
			}else{
				start.setXYZ(start.x,end.y * i,start.z);
				end.setXYZ(end.x,tmpstart.y * i,end.z);
			}
		}
	}
	private void shift(int x, int y, int z) {
		start.add(x,y,z);
		end.add(x,y,z);
	}
	private void rotateXZ(int d) {
		for(int count = 0;count < breaknum ; count++){
			Coordinate start = startlist.get(count);
			Coordinate end = endlist.get(count);
			Coordinate tmpstart = new Coordinate(startlist.get(count));
			switch(d){
			case 90:
				start.setXYZ(-end.z, start.y, start.x);
				end.setXYZ(-tmpstart.z,end.y,end.x);
				break;
			case 180:
				start.setZ(-end.z);
				end.setZ(-tmpstart.z);
				break;
			case 270:
				start.setXYZ(start.z,start.y,start.x);
				end.setXYZ(end.z, end.y, end.x);
				break;
			case 360:
				break;
			}
		}
	}
	public Coordinate getBreakLength() {
		return breaklength;
	}
	public int getBreakNum() {
		return breaknum;
	}


}

File: src/main/scala/com/github/unchama/seichiassist/data/BreakArea.scala
package com.github.unchama.seichiassist.data

import cats.effect.IO
import com.github.unchama.seichiassist.ActiveSkill
import com.github.unchama.seichiassist.util.BreakUtil
import org.bukkit.entity.Player

/**
 * 与えられたアクティブスキルの情報に基づいて、
 * プレーヤーがスキルで破壊する範囲を計算するデータオブジェクト
 * @param `type` スキルタイプ番号
 * @param level スキルレベル
 * @param mineflagnum フラグ
 * @param assaultflag アサルトスキルの時true
 */
class BreakArea(val `type`: Int,
                val level: Int,
                val mineflagnum: Int,
                val assaultflag: Boolean) {
  private val skill: ActiveSkill = ActiveSkill.values.apply(`type` - 1)

  //南向きを基準として破壊の範囲座標
  val breakLength: XYZTuple = skill.getBreakLength(level)
  //破壊回数
  val breakNum: Int = skill.getRepeatTimes(level)

  private def breakAreaListFromDirection(dir: String): List[AxisAlignedCuboid] = {
    import BreakArea.CoordinateManipulation._
    import syntax._

    val firstShift: AxisAlignedCuboid => AxisAlignedCuboid =
      if (assaultflag && `type` == 6 && level == 10) {
        //アサルトスキルの時
        areaShift(XYZTuple(0, (breakLength.y - 1) / 2 - 1, 0))
      } else if (dir == "U" || dir == "D" && (assaultflag || level >= 3)) {
        //上向きまたは下向きの時
        areaShift(XYZTuple(0, (breakLength.y - 1) / 2, 0))
      } else {
        //それ以外の範囲
        areaShift(XYZTuple(0, (breakLength.y - 1) / 2 - 1, (breakLength.z - 1) / 2))
      }

    val secondShift: AxisAlignedCuboid => AxisAlignedCuboid =
      if (`type` == ActiveSkill.BREAK.gettypenum && level < 3)
        incrementYOfEnd
      else
        identity

    val thirdShift: AxisAlignedCuboid => AxisAlignedCuboid =
      if (`type` == ActiveSkill.BREAK.gettypenum && level < 3 && mineflagnum == 1)
        areaShift(XYZTuple(0, 1, 0))
      else
        identity

    val directionalShift: AxisAlignedCuboid => AxisAlignedCuboid =
      dir match {
        case "N" | "E" | "S" | "W" =>
          areaShift(XYZTuple(0, 0, breakLength.z))
        case "U" | "D" if assaultflag || level >= 3 =>
          areaShift(XYZTuple(0, breakLength.y, 0))
        case _ => identity
      }

    val rotation: AxisAlignedCuboid => AxisAlignedCuboid =
      dir match {
        case "N" => rotateXZ(180)
        case "E" => rotateXZ(270)
        case "W" => rotateXZ(90)
        case "D" if !assaultflag => invertY
        // 横向きのスキル発動の場合Sが基準となり、
        // 縦向きの場合Uが基準となっているため回転しないで良い
        case "S" | "U" | _ => identity
      }

    val firstArea = {
      // 中心が(0,0,0)である領域(start = -end)を変形していく。
      val end = (breakLength - XYZTuple(1, 1, 1)) / 2.0
      val start = end.negative

      import scala.util.chaining._

      AxisAlignedCuboid(end, start)
        .pipe(firstShift)
        .pipe(secondShift)
        .pipe(thirdShift)
    }

    LazyList
      .iterate(firstArea)(directionalShift)
      .map(rotation)
      .take(breakNum)
      .toList
  }

  def makeBreakArea(player: Player): IO[List[AxisAlignedCuboid]] =
    BreakArea.getCardinalDirection(player).map(breakAreaListFromDirection)
}

object BreakArea {
  private val getCardinalDirection: Player => IO[String] = { player => IO { BreakUtil.getCardinalDirection(player) } }

  object CoordinateManipulation {
    import syntax._

    val incrementYOfEnd: AxisAlignedCuboid => AxisAlignedCuboid = {
      case area@AxisAlignedCuboid(_, end@XYZTuple(_, y, _)) =>
        area.copy(end = end.copy(y = y + 1))
    }

    val invertY: AxisAlignedCuboid => AxisAlignedCuboid = { case AxisAlignedCuboid(begin, end) =>
      def invertYOfVector(vector: XYZTuple): XYZTuple = XYZTuple(vector.x, -vector.y, vector.z)

      AxisAlignedCuboid(invertYOfVector(begin), invertYOfVector(end))
    }

    def areaShift(vector: XYZTuple): AxisAlignedCuboid => AxisAlignedCuboid = {
      case AxisAlignedCuboid(begin, end) =>
        AxisAlignedCuboid(begin + vector, end + vector)
    }

    def rotateXZ(d: Int): AxisAlignedCuboid => AxisAlignedCuboid = { case AxisAlignedCuboid(begin, end) =>
      d match {
        case 90 =>
          AxisAlignedCuboid(XYZTuple(-end.z, begin.y, begin.x), XYZTuple(-begin.z, end.y, end.x))
        case 180 =>
          AxisAlignedCuboid(XYZTuple(begin.x, begin.y, -end.z), XYZTuple(end.x, end.y, -begin.z))
        case 270 =>
          AxisAlignedCuboid(XYZTuple(begin.z, begin.y, begin.x), XYZTuple(end.z, end.y, end.x))
        case 360 =>
          AxisAlignedCuboid(begin, end)
      }
    }
  }
}
